contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,259 @@
+            Token token = tokens[i];
+            // skip token if no token balance or token is BNT - no need to swap in this case
+            if (balances[i] == 0 || token == _bnt) {
+                continue;
+            }
+
+            // get trade amount for token
+            uint256 tradeAmount = balances[i] - rewardAmounts[i];
+
+            // approve tokens for trading on Bancor Network V3
+            _setAllowance(token, tradeAmount);
+
+            uint256 val = token.isNative() ? tradeAmount : 0;
+
+            // swap token to BNT using Bancor Network V3
+            _bancorNetwork.tradeBySourceAmount{ value: val }(token, _bnt, tradeAmount, 1, block.timestamp, address(0));
+        }
+
+        // allocate rewards to caller and burn the rest
+        _allocateRewards(msg.sender, tokens, rewardAmounts);
+    }
+
+    /**
+     * @dev allocates the rewards to caller and burns the rest
+     */
+    function _allocateRewards(address sender, Token[] calldata tokens, uint256[] memory rewardAmounts) private {
+        // transfer the rewards to caller
+        uint256 len = tokens.length;
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            uint256 rewardAmount = rewardAmounts[i];
+            if (rewardAmount == 0) {
+                continue;
+            }
+            if (token.isNative()) {
+                // using a regular transfer here could revert due to exceeding the 2300 gas limit
+                // which is why we're using call instead (via sendValue)
+                payable(sender).sendValue(rewardAmount);
+            } else {
+                token.safeTransfer(sender, rewardAmount);
+            }
+        }
+
+        // get the burn amount
+        uint256 burnAmount = _bnt.balanceOf(address(this));
+
+        if (burnAmount > 0) {
+            // add to the total burned amount
+            _totalBurned += burnAmount;
+            // burn the tokens
+            _bnt.safeTransfer(Token.unwrap(_bnt), burnAmount);
+        }
+
+        emit FeesBurned(sender, tokens, rewardAmounts, burnAmount);
+    }
+
+    /**
+     * @dev set allowance to Bancor Network V3 to the max amount if it's less than the input amount
+     */
+    function _setAllowance(Token token, uint256 inputAmount) private {
+        if (token.isNative()) {
+            return;
+        }
+        uint256 allowance = token.allowance(address(this), address(_bancorNetwork));
+        if (allowance < inputAmount) {
+            // increase allowance to the max amount if allowance < inputAmount
+            token.safeIncreaseAllowance(address(_bancorNetwork), type(uint256).max - allowance);
+        }
+    }
+
+    function uncheckedInc(uint256 i) private pure returns (uint256 j) {
+        unchecked {
+            j = i + 1;
+        }
+    }
+
+    /**
+     * @dev perform various validations for the token array
+     */
+    modifier validateTokens(Token[] calldata tokens) {
```

Comment:
Nitpick> In the past, we have been usually defining modifiers at the top and moving its logic to a helper function.

Reply:
updated